graphql/src/Mu/GraphQL/Quasi.hs

```diff
@@ -0,0 +1,106 @@
+{-# language DataKinds         #-}
+{-# language OverloadedStrings #-}
+{-# language TemplateHaskell   #-}
+{-# language ViewPatterns      #-}
+
+module Mu.GraphQL.Quasi where
+
+import           Data.Int                      (Int32)
+import qualified Data.Text                     as T
+import           Data.UUID                     (UUID)
+import           Language.GraphQL.Draft.Parser (parseSchemaDoc)
+import qualified Language.GraphQL.Draft.Syntax as GQL
+import           Language.Haskell.TH
+
+import           Mu.GraphQL.Annotations
+import           Mu.Rpc
+import           Mu.Schema.Definition
+
+-- | Imports an graphql definition written in-line as a 'Schema'.
+graphql :: String -> String -> T.Text -> Q [Dec]
+graphql scName svName schema =
+  case parseSchemaDoc schema of
+    Left e  -> fail ("could not parse graphql spec: " ++ show e)
+    Right p -> graphqlToDecls scName svName p
+
+data Result =
+    GQLSchema Type
+  | GQLService Type
+  deriving (Show, Eq)
+
+-- | Constructs the GraphQL tree splitting between Schemas and Services.
+graphqlToDecls :: String -> String -> GQL.SchemaDocument -> Q [Dec]
+graphqlToDecls schemaName serviceName (GQL.SchemaDocument types) = do
+  let schemaName'  = mkName schemaName
+      serviceName' = mkName serviceName
+  rs <- traverse (typeToDec schemaName') types
+  let schemaTypes  = [x | GQLSchema  x <- rs]
+      serviceTypes = [x | GQLService x <- rs]
+  schemaDec <- tySynD schemaName' [] (pure $ typesToList schemaTypes)
+  serviceDec <- tySynD serviceName' [] (pure $ typesToList serviceTypes)
+  pure [schemaDec, serviceDec]
+
+-- | Reads a GraphQL 'TypeDefinition' and returns a 'Result'.
+typeToDec :: Name -> GQL.TypeDefinition -> Q Result
+typeToDec _ (GQL.TypeDefinitionScalar (GQL.ScalarTypeDefinition _ s _)) = GQLSchema <$> scalarToType s
```

Comment:
I don't understand what this is doing. Scalars do not go into either the schema or the service definition (you can return `Maybe Result` here and then use `catMaybes` or `mapMaybe` in `graphqlToDecs`).

What you generate here is more like the thing that would go into an argument or field definition.